cmd/kubeadm/app/util/runtime/runtime.go

```diff
@@ -179,3 +181,52 @@ func (runtime *DockerRuntime) ImageExists(image string) (bool, error) {
 	err := runtime.exec.Command("docker", "inspect", image).Run()
 	return err == nil, nil
 }
+
+// isExistingSocket checks if path exists and is domain socket
+func isExistingSocket(path string) bool {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return false
+	}
+
+	return fileInfo.Mode()&os.ModeSocket != 0
+}
+
+// detectCRISocketImpl is separated out only for test purposes, DON'T call it directly, use DetectCRISocket instead
+func detectCRISocketImpl(isSocket func(string) bool) (string, error) {
+	foundCRISockets := []string{}
+	knownCRISockets := []string{
+		// /var/run/dockershim.sock is deliberately omitted from here, we have a separate Docker check below
+		"/run/containerd/containerd.sock",
+		"/var/run/crio/crio.sock",
+	}
+
+	const dockerSocket = "/var/run/docker.sock" // The Docker socket is not CRI compatible
+	if isSocket(dockerSocket) {
+		foundCRISockets = append(foundCRISockets, kubeadmapiv1alpha3.DefaultDockerCRISocket)
+	}
+
+	for _, socket := range knownCRISockets {
+		if isSocket(socket) {
```

Comment:
Checking existence of the socket is not enough from my point of view. I'd suggest to also check if CRI API is accessible through socket.

Reply:
maybe we should do https://godoc.org/k8s.io/kubernetes/pkg/kubelet/apis/cri#RuntimeVersioner
or something that is know to be supported by all implementers.

Reply:
crictl info should be enough, I believe.

Reply:
I don't think, that we need to make the check too complex. Even checking, that the path leads to a domain socket is a bit of an overkill on my part. The sockets we are checking for at the moment, cannot be created without root privileges and I doubt that any process running as root could create an exact socket path by accident.

My opinion is to keep things as simple as possible. If there aren't any use cases, where someone could deliberately setup an invalid CRI socket, then I don't think we should add any additional checks for it.